flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -123,14 +124,12 @@ protected void performDefaultAction(ActionContext context) throws Exception {
 		context.allActionsCompleted();
 	}
 
-	private void runAlternativeMailboxLoop() throws InterruptedException {
+	private void runAlternativeMailboxLoop() throws InterruptedException, MailboxStateException {
 
-		while (true) {
+		final Mailbox mailbox = taskMailboxExecutor.getMailbox();
```

Comment:
Because you are using the mailbox here directly, you can also add a precondition similar to `taskMailboxExecutor.checkIsMailboxThread()`.

Otherwise, it's implicitly not compatible with some tests now. I've worked on a fix in the checkpoints-to-mailbox ticket, 
if you want, I can try to cherry pick them before your commit, to properly check that tests are good with your PR.

Reply:
@StefanRRichter, the patch for tests: #8858